ste/sender-pageBlobFromURL.go

```diff
@@ -33,16 +33,32 @@ import (
 type urlToPageBlobCopier struct {
 	pageBlobSenderBase
 
-	srcURL url.URL
+	srcURL      url.URL
+	srcPageList *azblob.PageList
 }
 
 func newURLToPageBlobCopier(jptm IJobPartTransferMgr, destination string, p pipeline.Pipeline, pacer pacer, srcInfoProvider IRemoteSourceInfoProvider) (s2sCopier, error) {
+	srcURL, err := srcInfoProvider.PreSignedSourceURL()
+	if err != nil {
+		return nil, err
+	}
 
 	destBlobTier := azblob.AccessTierNone
-	// If the source is page blob, preserve source's blob tier.
+	var srcPageList *azblob.PageList
 	if blobSrcInfoProvider, ok := srcInfoProvider.(IBlobSourceInfoProvider); ok {
 		if blobSrcInfoProvider.BlobType() == azblob.BlobPageBlob {
+			// if the source is page blob, preserve source's blob tier.
 			destBlobTier = blobSrcInfoProvider.BlobTier()
+
+			// also get the page ranges so that we can skip the empty parts at the expense of one HTTP request
+			srcPageBlobURL := azblob.NewPageBlobURL(*srcURL, p)
+			ctx := context.WithValue(jptm.Context(), ServiceAPIVersionOverride, azblob.ServiceVersion)
+
+			// ignore the error, since according to the REST API documentation:
```

Comment:
I really want this comment to be true (because its convenient) but I'm not sure that it actually is.  Why? Because of what I read here:
https://blogs.msdn.microsoft.com/windowsazurestorage/2012/03/26/getting-the-page-ranges-of-a-large-page-blob-in-segments/

I'm imaging, say, an 8 TB page blob that's been used as a VM disk.  It will have had many writes to it, but what if the disk was only 1/4 full.  So there were heaps of writes to the first 2 TB, but the remaining 6 TB is still all zeroes.  Maybe, those writes to the first 2 TB would be enough to cause us a timeout, leading us to incorrectly assume that the _whole blob_ is not sparse.

Maybe we could use the approach in the URL above, using a fairly coarse-grained approach (e.g. 1 TB at a time). That way, even if we error out on one TB, we won't make multi-TBs worth of wrong assumptions about the rest of the blob.

On one hand, given time constraints and deadlines, I'm not sure we should do this fix now. But on the other hand, if we release it, say its fixed, and folks are still getting TB's more used size than they expect, they won't be very pleased and it won't really be fixed for them.  Another option is, perhaps, to defer the whole PR until we have time to do the chunking, and another is to release it as-is, with the changelog saying that its a partial fix, with a fuller one to follow.

Reply:
That's some great insights! The article is from 7 years ago though, so I'll reach out to Michael Roberson to confirm whether the contents are still relevant and whether we should be mindful of the issue. 

We could also do tests to measure the current behavior, but that may be affected by how good of a day the service is having. Even if we get good results in our test, I would feel hesitant to rely on it as fact, so it's better to confirm with the service folks anyways.

I think in general it's good to ship incremental improvements. As long as we write a reasonable change log entry, I'm ok with releasing the current solution, and follow up with improvements if necessary. I see it as achieving a balance between saving cost vs performance vs complexity.

Reply:
Agreed.